Sources/Datadog/DatadogCore/Context/NetworkConnectionInfoPublisher.swift

```diff
@@ -0,0 +1,202 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+
+internal protocol NetworkConnectionInfoPublisher: ContextValuePublisher, ContextValueReader where Value == NetworkConnectionInfo? {
+    func set(queue: DispatchQueue) -> Self
+}
+
+internal struct AnyNetworkConnectionInfoPublisher: NetworkConnectionInfoPublisher {
+    private let publisher: AnyContextValuePublisher<NetworkConnectionInfo?>
+    private let reader: AnyContextValueReader<NetworkConnectionInfo?>
+    private let setQueue: (DispatchQueue) -> AnyNetworkConnectionInfoPublisher
+
+    init<Publisher>(_ publisher: Publisher) where Publisher: NetworkConnectionInfoPublisher {
+        self.publisher = publisher.eraseToAnyPublisher()
+        self.reader = publisher.eraseToAnyReader()
+        self.setQueue = { publisher.set(queue: $0).eraseToAnyPublisher() }
+    }
+
+    init() {
+        if #available(iOS 12, tvOS 12, *) {
+            self.init(NWPathMonitorPublisher())
+        } else {
+            self.init(SCNetworkReachabilityReader())
+        }
+    }
+
+    func set(queue: DispatchQueue) -> Self {
+        .init(setQueue(queue))
+    }
```

Comment:
This one is though, and it does not read well I admit.

> the `AnyNetworkConnectionInfoPublisher` now requires the caller to follow a convention of calling set(queue:) in order to make things work as expected. It's lacking type safety (that is typically guaranteed by accepting queue in init).

This issue is that the publisher is instantiated before the context provider queue is created, so we need a way of injecting the queue. I made it so `set(queue:)` is **not required**, and it will work on the `.global(qos: .utility)` queue if no queue is specified.

> Maybe if choosing `final class` over `struct` we can mutate the state in-place and replace queue from `init()`?

The difficulty here is that only `NWPathMonitorPublisher` needs a queue, not `SCNetworkReachabilityReader`, so we need to make it work for both. `AnyNetworkConnectionInfoPublisher` does not mutate itself, but it's underlying `NWPathMonitorPublisher` does.

I will see if I can improve this! Maybe `NWPathMonitorPublisher` can keep working on a the `.global(qos: .utility)` queue and does not need injection 🤔

Reply:
Ok, so I made the `DatadogContextProvider` agnostic to publisher and reader. We can now easily subscribe to a publisher, or assign a reader to a context value.

Also, publishers can run on different queue that the context's queue, we don't need to inject it anymore.

Reply:
New approach is much more straightforward, well done 👍.